linux/git_mirror

```diff
@@ -813,21 +848,181 @@ function git_push_main()
+#
+# .. seealso::
+#
+#   :func:`git_push_main`
+#
+# .. note::
+#
+#   Because this function creates a tag directory called just_git_mirror/ (note the trailing slash), a user can no longer create a tag called just_git_mirror
+#**
+
+function git_tag_main()
+{
+  if [ $# -eq 0 ]; then
+    echo "Usage:" >&2
+    echo "${BASH_SOURCE[0]} REPO_MAP_FILE PREP_DIR" >&2
+    exit 1
+  fi
+
+  # By default, only tag refs/heads/*
+  local GIT_MIRROR_TAG_REFS_REGEX="${GIT_MIRROR_TAG_REFS_REGEX-^refs/heads/.+}"
+  # REVIEW Should I have a GIT_MIRROR_SKIP_REFS_REGEX here too
+  # NOTE These regex's must be kept in sync with git_push_main
+  # By default, push all refs
+  local GIT_MIRROR_PUSH_REFS_REGEX="${GIT_MIRROR_PUSH_REFS_REGEX-.+}"
+  # This regex is backwards intentionally: if unset, don't exclude any refs
+  local GIT_MIRROR_SKIP_REFS_REGEX="${GIT_MIRROR_SKIP_REFS_REGEX-$^}"
+
+  _git_mirror_load_info "${1}" # Sets repo_paths and repo_urls
+
+  pushd "${2}" &> /dev/null
+    local tag_date="$(date -u '+%Y-%m-%dT%H.%M.%SZ')" # ISO8061; ':' isn't valid
+
+    local index
+    # Loop through the repos
+    for index in ${repo_paths[@]+"${!repo_paths[@]}"}; do
+      local repo_url="${repo_urls[$index]}"
+      local repo_path="$(shopt -s nullglob; dirname */"${repo_paths[$index]}/config")"
+      next_section "Tagging refs in '${repo_path}'"
+      if [ -d "${repo_path}" ]; then
+        pushd "${repo_path}" &> /dev/null
+          # refs take their long form to maintain namespacing of the tag;
+          # otherwise, refs/heads/4 and refs/pull/4 would both be tagged
+          # identically
+          #
+          # Spaces are not allowed in a refname
+          local refs=($("${GIT}" for-each-ref --format='%(refname)' | \
+              grep -E "${GIT_MIRROR_TAG_REFS_REGEX}" | \
+              grep -E "${GIT_MIRROR_PUSH_REFS_REGEX}" | \
+              grep -v -E "${GIT_MIRROR_SKIP_REFS_REGEX}" | \
+              grep -v -e '^refs/tags/just_git_mirror/' -e '^refs/remotes' -e '^refs/stash' || :))
+          for ref in ${refs+"${refs[@]}"}; do
+            # Don't re-tag a ref if we have already tagged it at this SHA.
+            # Unfortunately, it doesn't look like there is something like
+            # git tag --points-at for ls-remote. Instead, return all of our
+            # tags that match our ref and grep for the correct SHA
+            #
+            # git has two kinds of tags: lightweight and annotated. Annotated
+            # tags are commits of their own [1]. In order to list the commit
+            # referred to by the tag (instead of the tag's commit itself) with
+            # git show-ref (from a local repo), use the --dereference flag, or
+            # with git rev-parse (from a local repo) or ls-remote (from a
+            # remote repo), peel the ref with ^{} [2], e.g.,
+            # git rev-pasre refs/tags/atag^{}
+            # We are creating annotated tags (because including a message implies
+            # annotation)
```

Comment:
> Annotated tags are meant for release while lightweight tags are meant for private or temporary object labels. For this reason, some git commands for naming objects (like git describe) will ignore lightweight tags by default.

[1] - https://git-scm.com/docs/git-tag

You really shouldn't be misusing annotated tags like that.

Reply:
True, but git also says this

> "It’s generally recommended that you create annotated tags so you can have all this information; but if you want a temporary tag or for some reason don’t want to keep the other information, lightweight tags are available too.


https://git-scm.com/book/en/v2/Git-Basics-Tagging